app/src/main/java/org/thoughtcrime/securesms/mms/DecryptableStreamLocalUriFetcher.java

```diff
@@ -0,0 +1,51 @@
+package org.thoughtcrime.securesms.mms;
+
+import android.content.ContentResolver;
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.net.Uri;
+
+import com.bumptech.glide.load.data.StreamLocalUriFetcher;
+
+import org.signal.core.util.logging.Log;
+import org.thoughtcrime.securesms.util.MediaUtil;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+
+class DecryptableStreamLocalUriFetcher extends StreamLocalUriFetcher {
+
+  private static final String TAG = Log.tag(DecryptableStreamLocalUriFetcher.class);
+
+  private Context context;
+
+  DecryptableStreamLocalUriFetcher(Context context, Uri uri) {
+    super(context.getContentResolver(), uri);
+    this.context = context;
+  }
+
+  @Override
+  protected InputStream loadResource(Uri uri, ContentResolver contentResolver) throws FileNotFoundException {
+    if (MediaUtil.hasVideoThumbnail(context, uri)) {
+      Bitmap thumbnail = MediaUtil.getVideoThumbnail(context, uri, 1000);
+
+      if (thumbnail != null) {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        thumbnail.compress(Bitmap.CompressFormat.JPEG, 100, baos);
+        ByteArrayInputStream thumbnailStream = new ByteArrayInputStream(baos.toByteArray());
+        thumbnail.recycle();
+        return thumbnailStream;
+      }
+    }
+
+    try {
+      return PartAuthority.getAttachmentThumbnailStream(context, uri);
+    } catch (IOException ioe) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)